pkgs/development/php-packages/php-parallel-lint/default.nix

```diff
@@ -1,40 +1,31 @@
-{ mkDerivation, fetchFromGitHub, makeWrapper, lib, php }:
+{ mkDerivation, fetchurl, makeWrapper, lib, php }:
 let
   pname = "php-parallel-lint";
-  version = "1.0.0";
+  version = "1.3.1";
 in
 mkDerivation {
   inherit pname version;
 
-  src = fetchFromGitHub {
-    owner = "JakubOnderka";
-    repo = "PHP-Parallel-Lint";
-    rev = "v${version}";
-    sha256 = "16nv8yyk2z3l213dg067l6di4pigg5rd8yswr5xgd18jwbys2vnw";
+  src = fetchurl {
+    url = "https://github.com/php-parallel-lint/PHP-Parallel-Lint/releases/download/v${version}/parallel-lint.phar";
+    sha256 = "sha256-vdzkegZ3roomRH2edaymBq34G9Jm2dCo6SBgBMoLL4I=";
   };
 
-  nativeBuildInputs = [
-    makeWrapper
-    php.packages.composer
-    php.packages.box
-  ];
+  dontUnpack = true;
 
-  buildPhase = ''
-    composer dump-autoload
-    box build
-  '';
+  nativeBuildInputs = [ makeWrapper ];
```

Comment:
I mostly do not agree and I will make my point, perhaps this is something I don't understand yet.

First, I don't think that these PHP projects should be packaged in Nix. I was surprised to see that they were available at first.
Why? They are available through `composer` and as a `phar`.

Second, let's say that we decide to bundle this package in Nix, then I don't think it's up to us to build it.
There is a phar provided by the vendor, we should use it.

Third, the issue you're mentioning is in your project, not in Nix, so, maybe we should mitigate and fix the issue you had in your package.

Last but not least, in order to unify and simplify things, it's always nice to have the same building procedure for all projects.
Then, if we decide to build this PHP project, then we should also decide to do it for all the others that are in nixpkgs (phpstan, psalm, etc etc).

Anyway. I'm in this community since not so long ago so my voice would not weight in the balance, but this is what I think as a PHP developer.

Reply:
> First, I don't think that these PHP projects should be packaged in Nix. I was surprised to see that they were available at first. Why? They are available through `composer` and as a `phar`.

- Nix has much better ergonomics than composer:
  - there is no `nix-shell` equivalent
  - depending on who you ask, you are supposed to install projects that you do not want to or cannot (e.g. due to conflicting dependencies) install into your project using `composer create-project` or `composer global require`, which are both broken.
- phars are better but how are you supposed to use them?
  - downloading manually makes updating cumbersome
  - what is even [phive](https://phar.io/)?
  - Nix is the least bad package manager there is so we might as well use it
 
> Second, let's say that we decide to bundle this package in Nix, then I don't think it's up to us to build it. There is a phar provided by the vendor, we should use it.

As you noticed, there are two somewhat conflicting goals. Nix is source-based package manager and we prefer building packages from source where practical, since it allow us to fix various issues trivially (bugs that need patching, dependencies that need updating). But we also want to have the packages as close to upstream as possible to minimize the risk of weirdness. If we took the latter to conclusion, we would only use binary packages provided by upstream. But that would go against the first goal, which is IMO more important for a Linux distribution.

> Third, the issue you're mentioning is in your project, not in Nix, so, maybe we should mitigate and fix the issue you had in your package.

Yes, in the end it turned out that it was issue in our PHP packaging but workaround would still require rebuilding the phar ourselves. Issues that require patching the source or tweaking the build process are bread and butter of package maintainer. 

> Last but not least, in order to unify and simplify things, it's always nice to have the same building procedure for all projects. Then, if we decide to build this PHP project, then we should also decide to do it for all the others that are in nixpkgs (phpstan, psalm, etc etc).

Yes unification is a nice thing to strive for but different packages have different maintainers and it is not really fair to expect a person who wants to have a package available to also normalize all other packages they might not even use.

> Anyway. I'm in this community since not so long ago so my voice would not weight in the balance, but this is what I think as a PHP developer.

I think fresh perspective is always useful since we might accustomed to the status quo without realizing that it is suboptimal. But it is also important to realize that each custom usually has a reason behind it.

Reply:
>     * Nix has much better ergonomics than composer:
>       
>       * there is no `nix-shell` equivalent
>       * depending on who you ask, you are supposed to install projects that you do not want to or cannot (e.g. due to conflicting dependencies) install into your project using `composer create-project` or `composer global require`, which are both broken.
> 
>     * phars are better but how are you supposed to use them?
>       
>       * downloading manually makes updating cumbersome
>       * what is even [phive](https://phar.io/)?
>       * Nix is the least bad package manager there is so we might as well use it

I'm not comparing `composer` and Nix. Those are 2 completely different things for me, no discussion.
What I meant is that, when I need `php-parallel-lint` in [a project](https://packagist.org/packages/drupol/php-conventions), I add it myself in `require-dev` and do not rely on the underlying OS to provide it for me. And I still believe that these packages does not have their place in Nix... but that's another subject.

"phars are better" - let me explain.
composer, phpstan, psalm... are all providing PHARS that are built out of sources in Github Actions.
Millions of people are relying on these phars already and there unlikely few chances that it would broke one day.
Even the default `composer` [installer](https://getcomposer.org/download/) is using the phar version.

Another example, when PHP developers are requiring `phpstan/phpstan` in any projects, they are actually downloading the PHAR version of the project, not the PHP sources. This has the advantages of making sure that it works and there won't be any dependency conflicts. The phar version of PHPStan embed namespaced dependencies only required by PHPStan and does not interfere with the local project. The PHAR building is completely automatized on Github actions as well.

> As you noticed, there are two somewhat conflicting goals. Nix is source-based package manager and we prefer building packages from source where practical, since it allow us to fix various issues trivially (bugs that need patching, dependencies that need updating). But we also want to have the packages as close to upstream as possible to minimize the risk of weirdness. If we took the latter to conclusion, we would only use binary packages provided by upstream. But that would go against the first goal, which is IMO more important for a Linux distribution.

I understand. Then maybe in that case, we should provide 2 packages ?
* `php-parallel-lint` would be the version compiled from sources
* `php-parallel-lint-bin` would be the PHAR version.

I'm all in favor of open-source but I have the feeling that we are redoing twice(or more) the job here and this is what I'm trying to avoid at all costs (_for many reasons that I would not expose here but mostly environmental_). If the vendor is providing the PHAR, automatically built on Github actions, we should use it. That's usually the default way of using such packages anyway.

> Yes unification is a nice thing to strive for but different packages have different maintainers and it is not really fair to expect a person who wants to have a package available to also normalize all other packages they might not even use.

About `php-parallel-lint`, I think it worth to mention that I'm updating this package from version 1.0.0 (2018) to version 1.3.1 (2021) and things has changed a lot. Maybe we should try to use the PHAR for the time being so it's aligned with the rest of phar packages in use, and see how it behaves? Let's give it a try and as soon as there is an issue, I'll be the first to submit the PR to built this out of the source.

> I think fresh perspective is always useful since we might accustomed to the status quo without realizing that it is suboptimal. But it is also important to realize that each custom usually has a reason behind it.

Thanks, I was a bit afraid that this message was misunderstood, and I'm happy to see that it is a constructive discussion in the end, let's keep it like that.

Reply:
> I'm not comparing `composer` and Nix. Those are 2 completely different things for me, no discussion. What I meant is that, when I need `php-parallel-lint` in [a project](https://packagist.org/packages/drupol/php-conventions), I add it myself in `require-dev` and do not rely on the underlying OS to provide it for me. And I still believe that these packages does not have their place in Nix... but that's another subject.

Well sometimes you cannot add a project to `require-dev` (for example because of conflicting dependencies) or you do not want to do that (I regularly run `nix shell -f /home/jtojnar/Projects/nix-phps outputs.packages.x86_64-linux.php56 outputs.packages.x86_64-linux.php56.packages.php-parallel-lint -c php-parallel-lint src` when contributing to third party libraries.)

> "phars are better" - let me explain. composer, phpstan, psalm... are all providing PHARS that are built out of sources in Github Actions. Millions of people are relying on these phars already and there unlikely few chances that it would broke one day. Even the default `composer` [installer](https://getcomposer.org/download/) is using the phar version.

Well, you could say the same thing for binaries. Though it is true that this is a less of an issue here because PHP projects’ build process is less complex and there are no ABI boundaries.

> Another example, when PHP developers are requiring `phpstan/phpstan` in any projects, they are actually downloading the PHAR version of the project, not the PHP sources. This has the advantages of making sure that it works and there won't be any dependency conflicts. The phar version of PHPStan embed namespaced dependencies only required by PHPStan and does not interfere with the local project. The PHAR building is completely automatized on Github actions as well.

I am aware of that but that is still pretty rare for Packagist packages. And not without downsides either – particularly, it makes debugging issues with the tool hard.

> > As you noticed, there are two somewhat conflicting goals. Nix is source-based package manager and we prefer building packages from source where practical, since it allow us to fix various issues trivially (bugs that need patching, dependencies that need updating). But we also want to have the packages as close to upstream as possible to minimize the risk of weirdness. If we took the latter to conclusion, we would only use binary packages provided by upstream. But that would go against the first goal, which is IMO more important for a Linux distribution.
> 
> I understand. Then maybe in that case, we should provide 2 packages ?
> 
>     * `php-parallel-lint` would be the version compiled from sources
> 
>     * `php-parallel-lint-bin` would be the PHAR version.

Providing both versions is also a possibility but it brings costs of its own – now consumers will need to decide which version they want to use. So it IMO only makes sense when the two versions differ significantly (e.g. `firefox-bin` is more performant but `firefox` supports declarative customization).

> I'm all in favor of open-source but I have the feeling that we are redoing twice(or more) the job here and this is what I'm trying to avoid at all costs (_for many reasons that I would not expose here but mostly environmental_). If the vendor is providing the PHAR, automatically built on Github actions, we should use it. That's usually the default way of using such packages anyway.

Well, that is precisely what I am not sure about. The current Nixpkgs policy is to build stuff from source unless too inconvenient. I would say the primary reason most projects used PHARs from upstream instead of building them from scratch is that it is not easy to do without IFD. But we should strive to respect the policy when it is not a great PITA. Particularly, php-parallel lint has no other runtime dependencies than PHP itself.

The environmental impact will be almost certainly overshadowed by every other part of the build since building phar is mostly just a compression.

Reply:
Thanks for taking the time to carefully explain each points.

I still do not agree with some of them, but I'm here to learn and I accept our disagreement.
The goal here is to move forward anyway :)

I did the changes you suggest and I hope it's ok now.